### PR TITLE
Project Plugins: Add Plugin.swift manifest.

### PR DESCRIPTION
**Pull request by luispadron**
## Plugin

A `Plugin` is defined as a directory that contains a `Plugin.swift` manifest file. This `Plugin.swift` must define a `plugin` that may look something like this:

```swift
// Plugin.swift

import ProjectDescription

let plugin = Plugin(name: "MyPlugin")
```

The list of supported `Plugin`s includes:

- `helper`: A plugin intended to extend `ProjectDescription`
  - These are independently compilable Swift frameworks i.e. they may only depend on `ProjectDescription`. 
  - Cannot be used in `Plugin.swift` or `Config.swift`.
 
### Implementation

`Plugin`s will be loaded before other manifests so that they may be used in other manifests such as `Project.swit`. They may then be injected and used for project generation or as in the use case of helpers, as a means of creating importable frameworks for usage within Tuist.

Because of issues that would be created by allowing this, importing `ProjectDescriptionHelpers` inside of `Plugin.swift` is disallowed. One such issue would be that becuase `ProjectDescriptionHelpers` can import plugins we would be creating a circular dependency between these two.

#### Future

- Shared templates for `tuist scaffold` (I'd love to see this, if I can find the time for it, I might try this myself) @fortmarek
- Defining custom commands that tuist can trigger - similar to when you run custom actions via Fastlane. @fortmarek
- We could also add a feature to add custom Xcode templates that could be shared with your team (we use fastlane for this [right now](https://github.com/AckeeCZ/ios-templates)) @fortmare
- Plugins could be mappers which modify the project graph in some way allowing for community driven support.

#### Side note

Would love to see some way for us to parse repos with this manifest and make them indexable, this could really help drive more usage of the feature.

### Implementation 👩‍💻👨‍💻

This is only the public API addition of Project Plugins, implementation is currently living in #2026 for helper plugins.

### CODE REVIEW

File: Sources/ProjectDescription/Config.swift [hunk 1]
```diff
@@ -125,29 +118,3 @@ extension Config.GenerationOptions {
...
-    case (.disableSynthesizedResourceAccessors, .disableSynthesizedResourceAccessors):
-        return true
-    case (.disableShowEnvironmentVarsInScriptPhases, .disableShowEnvironmentVarsInScriptPhases):
-        return true
-    case (.enableCodeCoverage, .enableCodeCoverage):
-        return true
-    default:
-        return false
-    }
-}
```
[the remark this answers is not in the document]
Comment (fortmarek) [reply Sources/ProjectDescription/Config.swift:153]: yeah, I think it should be fine to delete this

File: Sources/ProjectDescription/Plugins.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+/// A `Plugin` manifest allows for defining extensions for Tuist.
+///
+/// Supported plugins include:
+/// - ProjectDescriptionHelpers
+///     - These are plugins designed to be usable by any other manifest excluding `Config` and `Plugin`.
+///     - The source files for these helpers must live under a ProjectDescriptionHelpers directory in the location where `Plugin` manifest lives.
+///
+public struct Plugin: Codable, Equatable {
```
Comment (fortmarek) [inline Sources/ProjectDescription/Plugins.swift:10]: Nitpick: I'd rename the file to `Plugin.swift`, so the names correspond to each other

**Review by fortmarek — APPROVED** [review]
Thanks for making the PRs more digestible 🙂 just a small nitpick, otherwise this looks good

File: Sources/ProjectDescription/Plugins.swift (hunk 2 above)
Comment (luispadron) [reply Sources/ProjectDescription/Plugins.swift:10]: Oops nice catch

**Review by pepibumur — APPROVED** [review]

**Comment by pepibumur** [comment]
@all-contributors add @luispadron for code

**Pull request merged**